### PR TITLE
Added mTLS support to MimirClient

### LINKED ISSUE #426: [Feature Request]: Support client ceritifacte authentication for Mimir
### Feature Request
Add support for client ceritifacte authenticaiton to Mimir via cli
`grr config set mimir.tls.client-certificate-path /path/to/cert`
`grr config set mimir.tls.client-key-path /path/to/key`

We want to utilize this tool in our environment, we provide Mimir behind a gateway secured with client certificates.

### PR DESCRIPTION
**Pull request by Itaykal**
Added optional mTLS config to Mimir client.
Added an option to append an additional ca bundle file to mimir client.

I think it should wait for #415 as the changes could not be tested at the moment.

Fixes #426

### CODE REVIEW

File: pkg/config/model.go [hunk 1]
```diff
@@ -9,9 +9,16 @@ type GrafanaConfig struct {
 }
 
 type MimirConfig struct {
-	Address  string `yaml:"address" mapstructure:"address"`
-	TenantID string `yaml:"tenant-id" mapstructure:"tenant-id"`
-	APIKey   string `yaml:"api-key" mapstructure:"api-key"`
+	Address  string         `yaml:"address" mapstructure:"address"`
+	TenantID string         `yaml:"tenant-id" mapstructure:"tenant-id"`
+	APIKey   string         `yaml:"api-key" mapstructure:"api-key"`
+	TLS      MimirTLSConfig `yaml:"tls" mapstructure:"auth"`
```
Comment (malcolmholmes) [inline pkg/config/model.go:15]: Is there a reason for yaml being `tls` and mapstructure being `auth`?

File: pkg/config/model.go [hunk 2]
```diff
@@ -9,9 +9,16 @@ type GrafanaConfig struct {
...
+}
+
+type MimirTLSConfig struct {
+	ClientCertPath string `yaml:"client-cert-path,omitempty" mapstructure:"client-cert-path"`
+	ClientKeyPath  string `yaml:"client-key-path,omitempty" mapstructure:"client-key-path"`
+	CAPath 		   string `yaml:"ca-path" mapstruceture:"ca-path"`
```
Comment (malcolmholmes) [inline pkg/config/model.go:21]:
```suggestion
	CAPath 		   string `yaml:"ca-path" mapstructure:"ca-path"`
```

**Review by malcolmholmes — COMMENTED** [review]
This is an amazing contribution, thanks! I've left a couple of comments, but will also await comment from @spinillos as this touches her code.

File: pkg/config/model.go (hunk 1 above)
Comment (Itaykal) [reply pkg/config/model.go:15]: I forgot to change it :) fixing

File: pkg/mimir/client/http_client.go [hunk 3]
```diff
@@ -126,5 +128,42 @@ func createHTTPClient() (*http.Client, error) {
...
+		}
+
+		caCertPEM, err := os.ReadFile(c.config.TLS.CAPath)
+		if err != nil {
+			return nil, err
+		} 
+		
+		ok := certPool.AppendCertsFromPEM(caCertPEM)
+		if !ok {
+			return nil, err
```
Comment (spinillos) [inline pkg/mimir/client/http_client.go:151]: Error here is nil, you should return a custom error.

File: pkg/mimir/client/http_client.go [hunk 4]
```diff
@@ -126,5 +128,42 @@ func createHTTPClient() (*http.Client, error) {
...
+		}
+		
+		tlsConfig.RootCAs = certPool
+	}
+
+	if c.config.TLS.ClientCertPath != "" || c.config.TLS.ClientKeyPath != "" {
```
Comment (spinillos) [inline pkg/mimir/client/http_client.go:157]: Does it work if we only set one of them? In that case, should it be `&&`?

File: pkg/config/model.go [hunk 5]
```diff
@@ -9,9 +9,16 @@ type GrafanaConfig struct {
...
+	Address  string         `yaml:"address" mapstructure:"address"`
+	TenantID string         `yaml:"tenant-id" mapstructure:"tenant-id"`
+	APIKey   string         `yaml:"api-key" mapstructure:"api-key"`
+	TLS      MimirTLSConfig `yaml:"tls" mapstructure:"tls"`
+}
+
+type MimirTLSConfig struct {
+	ClientCertPath string `yaml:"client-cert-path,omitempty" mapstructure:"client-cert-path"`
+	ClientKeyPath  string `yaml:"client-key-path,omitempty" mapstructure:"client-key-path"`
+	CAPath 		   string `yaml:"ca-path" mapstructure:"ca-path"`
```
Comment (spinillos) [inline pkg/config/model.go:21]: It isn't formatted.

**Review by spinillos — COMMENTED** [review]
Thanks for your contribution 😃!! I added a few comments.

File: pkg/mimir/client/http_client.go (hunk 4 above)
Comment (Itaykal) [reply pkg/mimir/client/http_client.go:157]: `LoadX509KeyPair` should return an error if the file path doesn't exist and if the contents are not a valid certificate, is this enough or should I return an error when only one of them is configured?

Comment (spinillos) [reply pkg/mimir/client/http_client.go:157]: I think that it's enough. Thanks for the clarification.

**Review by spinillos — APPROVED** [review]

**Pull request merged**